File: microservices/product-services/src/models/books.rs
use mongodb::{bson::{doc, Document}, Collection, Database, options::FindOptions, error::Error};
use futures::TryStreamExt;

pub async fn get_books_count(db: Database) -> Result<u64, Error> {
    let collection: Collection<Document> = db.collection("books");
    let count = collection.count_documents(doc! {}, None).await?;
    Ok(count)
}

pub async fn get_books(db: Database, limit: i64, offset: u64, is_featured: bool) -> Result<Vec<Document>, Error> {
    println!("limit: {}, offset: {}, is_featured: {}", limit, offset, is_featured);
    let limited = if limit <= 0 || limit > 100 { 100 } else { limit};
    let options = FindOptions::builder()
                  .limit(Some(limited))
                  .skip(offset)
                  .build();
    
    let mut query = doc! {};

    if is_featured {
        query.insert("is_featured", true);
    }

    let collection: Collection<Document> = db.collection("books");    
    let mut cursor = collection.find(query, options).await?;
    let mut books: Vec<Document> = Vec::new();

    while let Some(book) = cursor.try_next().await? {
        books.push(book);
    }

    Ok(books)   
}

pub async fn get_grouped_books(db: Database) -> Result<Vec<Document>, Error> {
    let pipeline: Vec<Document> = vec![
        doc! { "$unwind": "$categories" },
        doc! {
            "$group": {
                "_id": "$categories",
                "books": {
                    "$push": {
                        "id": { "$toString": "$_id" },
                        "title": "$title",
                        "isbn": "$isbn",
                        "status": "$status",
                        "is_featured": "$is_featured",
                        "thumbnail_url": "$thumbnail_url",
                        // Include other fields as needed
                    }
                },
                "count": { "$sum": 1 }
            }
        },
        doc! { "$match": { "count": { "$gte": 5 } } },
        doc! { "$project": { "_id": 0, "category": "$_id", "books": { "$slice": ["$books", 10] } } },
    ];

    let mut grouped_books_cursor = db.collection::<Document>("books")
        .aggregate(pipeline, None)
        .await?;
    let mut grouped_books: Vec<Document> = Vec::new();

    while let Some(document) = grouped_books_cursor.try_next().await? {
        grouped_books.push(document);
    }

    Ok(grouped_books)
}

pub async fn get_book_detail(db: Database, book_id: String) -> Result<Document, Error> {
    let query = doc! {"isbn": book_id};

    let collection: Collection<Document> = db.collection("books");    
    let book_detail: Document = collection.find_one(query, None).await?.unwrap_or_default();

    Ok(book_detail)
}

File: monolithic/src/lib/books.ts
'use server'

import clientPromise from "../lib/mongodb";

export async function fetchBooks(limit: number, offset: number = 0, featured: boolean = false, category: string = "", search: string = "") {
    try {

        const client = await clientPromise;
        const db = client.db(process.env.MONGODB_NAME);
        const query: any = {};
        if (featured) {
            query.featured = featured;
        }
        if (category && category !== "") {
            query.categories = category;
        }
        if (search && search !== "") {
            query.$or = [
                { title: { $regex: search, $options: 'i' } }, 
                { authors: { $elemMatch: { $regex: search, $options: 'i' } } } 
            ];
        }
        const options = { 
            projection: {
                id: { $toString: "$_id" },
                title: 1,
                isbn: 1,
                status: 1,
                is_featured: 1,
                thumbnail_url: 1,
                price: 1,
                _id: 0
            },
            skip: offset, limit: limit };
        const booksData = await db.collection("books").find(query, options).toArray();

        return { results: booksData };
    } catch (error) {
        console.log('error', error);
        return { data: null, error: 'Failed to fetch book records: ' + error };
    }
}

export async function fetchBookDetails(bookId: string) {
    try {
        const client = await clientPromise;
        const db = client.db(process.env.MONGODB_NAME);
        const query: any = {};
        if (bookId != "" && bookId != null) {
            query.isbn = bookId;
        }
        const options = { 
            projection: {
                id: { $toString: "$_id" },
                title: 1,
                isbn: 1,
                status: 1,
                is_featured: 1,
                thumbnail_url: 1,
                short_description: 1,
                long_description: 1,
                authors: 1,
                categories: 1,
                price: 1,
                page_count: 1,
                _id: 0
            } };
        const bookData = await db.collection("books").findOne(query, options);

        return { results: bookData };
    } catch (error) {
        console.log('error', error);
        return { data: null, error: 'Failed to fetch book records: ' + error };
    }
}

export async function fetchGroupedBooks() {
    try {
        const client = await clientPromise;
        const db = client.db(process.env.MONGODB_NAME);

        const pipeline = [
            { $unwind: "$categories" }, 
            {
                $group: {
                    _id: "$categories", books: {
                        $push: {
                            id: { $toString: "$_id" },
                            title: "$title",
                            isbn: "$isbn",
                            status: "$status",
                            is_featured: "$is_featured",
                            thumbnail_url: "$thumbnail_url",
                            price: "$price"
                        }
                    }, count: { $sum: 1 } 
                }
            },
            {
                $match: { count: { $gte: 5 } } 
            }, 
            { $project: { _id: 0, category: "$_id", books: { $slice: ["$books", 10] } } } 
        ];

        const groupedBooks = await db.collection("books").aggregate(pipeline).toArray();

        return { results: groupedBooks };
    } catch (error) {
        console.log('error', error);
        return { data: null, error: 'Failed to fetch grouped book records: ' + error };
    }
}
